mbtools/copy_resources_s3.py

```diff
@@ -0,0 +1,137 @@
+    s3_client = boto3.client("s3")
+    try:
+        all_s3_resources = s3_client.list_objects(
+            Bucket=bucket,
+            Prefix=s3_dir
+        )
+        sha1_list = []
+        print(all_s3_resources)
+        for file_info in all_s3_resources['Contents']:
+            # extract the sha1 from the key
+            sha1 = file_info['Key'].split('/')[-1].split('.')[0]
+            sha1_list.append(sha1)
+    except botocore.exceptions.ClientError:
+        sha1_list = None
+    return sha1_list
+
+
+def new_resource_hashes(resource_dir):
+    """Generates sha1 hashes for all the resources in a local directory"""
+    sha1_map = {}
+    buff_size = 8 * 1024 * 1024
+    try:
+        for filename in os.listdir(resource_dir):
+
+            sha1 = hashlib.sha1()
+            full_path = os.path.join(resource_dir, filename)
+            with open(full_path, 'rb') as f:
+                while True:
+                    data = f.read(buff_size)
+                    if not data:
+                        break
+                    sha1.update(data)
+            sha1_map[sha1.hexdigest()] = full_path
+        return sha1_map
+    except FileNotFoundError:
+        return FileNotFoundError
+
+
+def compare_and_remove_hashes(keys, existing_keys):
+    """Returns a list of the items from keys that arent in existing_keys"""
+    new_keys = []
+    for i in keys:
+        if i not in existing_keys:
+            new_keys.append(i)
+    return new_keys
+
+
+def get_mime_type(filename):
+    """ get MIME type of file with libmagic """
+    mime_type = ''
+    try:
+        mime_type = magic.from_file(filename, mime=True)
+    finally:
+        return mime_type
+
+
+def add_new_resources_to_s3(bucket, s3_dir, hashes, filename_map,
+                            metadata_dir
+                            ):
+    """Add the files specified in filename_map to s3 iff they their
+    corresponding sha1 key exists in hashes"""
+    s3_client = boto3.client("s3")
+    print(hashes)
+    for key in hashes:
+        full_keypath = s3_dir + key
+        s3_client.upload_file(filename_map[key], bucket,
+                              full_keypath
+                              )
+        mime_type = get_mime_type(filename_map[key])
+        print(mime_type)
+        add_metadata(key, filename_map, metadata_dir, mime_type, full_keypath)
+
+
+def add_metadata(key, filename_map, metadata_dir, mime_type, full_keypath):
+    """"""
+    tag = {"mime_type": mime_type,
+           "sha1": key,
+           "original_filename": os.path.basename(filename_map[key]),
+           "s3_key": full_keypath
+           }
+
+    with open(metadata_dir) as f:
+        data = json.load(f)
+        data.append(tag)
+    with open(metadata_dir, 'w') as f:
+        json.dump(data, f)
+
+
+def main():
+    new_resource_dir = Path(sys.argv[1]).resolve(strict=True)
+    metadata_dir = Path(sys.argv[2]).resolve(strict=True)
```

Comment:
I was hoping to implement it such that this is a directory, and one can specify a directory of content to be uploaded all at once. Do you think specifying only a path to a specific file would be more clear/extensible?

Reply:
Oh, I think the directory is the right approach for input files, but that's the `new_resource_dir` input. The `metadata_dir` is the JSON file where you're adding elements `{"mime_type": mime_type, "sha1": key, "original_filename": os.path.basename(filename_map[key]), "s3_key": full_keypath }`, no?

Also it occurs to me now that this code assumes that file exists already on line 129. It might be nice to create it if it doesn't to avoid having to manually create it :man_shrugging: .